### ISSUE TITLE
Pass In Subaccount Details To Resources

### ISSUE THREAD

**Issue by enchorb** [RJPearson94/terraform-provider-twilio#54]
Hey, 

Instead of using an aliased provider for a subuser, would it be possible to add optional `account_sid` and `auth_token` properties on resources which could fall back to the provider ones? 

Current Way:
```
provider "twilio" {
  auth_token  = var.auth_token
  account_sid = var.account_sid
}

resource "twilio_account_sub_account" "sub_account" {
  friendly_name = var.name
}

provider "twilio" {
  alias        = "sub"
  auth_token   = local.first_apply ? var.auth_token : twilio_account_sub_account.sub_account.auth_token
  account_sid  = local.first_apply ? var.account_sid : twilio_account_sub_account.sub_account.sid
}

resource "twilio_verify_service" "verify" {
  provider       = twilio.sub
  friendly_name  = "Test Name"
  lookup_enabled = true
  skip_sms_to_landlines = true
  totp {
    issuer      =  "Test Name"
    time_step   = 30
    code_length = 6
    skew        = 1
  }
}
```

Proposed Way: 
Current Way:
```
provider "twilio" {
  auth_token  = var.auth_token
  account_sid = var.account_sid
}

resource "twilio_account_sub_account" "sub_account" {
  friendly_name = var.name
}

resource "twilio_verify_service" "verify" {
  auth_token   = twilio_account_sub_account.sub_account.auth_token
  account_sid  = twilio_account_sub_account.sub_account.sid
  friendly_name  = "Test Name"
  lookup_enabled = true
  skip_sms_to_landlines = true
  totp {
    issuer      =  "Test Name"
    time_step   = 30
    code_length = 6
    skew        = 1
  }
}
```

**Comment by RJPearson94** [comment]
Hi @enchorb, thanks for the feature request. I will have to do some research on this, as I think it will be tricky with the current implementation. I also want to see how other providers handle this, as I have tried to follow known/ expected patterns with this provider. 
I am curious to understand your use case a little more, I see you have a first apply local variable, can you provide some more info on why this is required I.e. is it to work around a limitation?

**Comment by enchorb** [comment]
Hey, yes the limitation is the on the first apply we get an error since the sub account hasn't been created yet when the aliased provider tries to authenticate to it  we get back an `[SID is required, Auth token is required]`. That's why have it as an optional argument on the resource would be useful but totally understand if it's against Terraform design patterns and can't be implemented.

**Comment by enchorb** [comment]
In case anyone needs this this shell script should help:

```
#!/bin/sh

PARENT_SID=$1
PARENT_TOKEN=$2
ACC_NAME=$3

main() {
  # Switch To Directory
  cd "$(dirname "$0")" || exit 1

  ACCOUNT=$(curl -X GET "https://api.twilio.com/2010-04-01/Accounts.json?FriendlyName=$ACC_NAME&PageSize=1" -u "$PARENT_SID:$PARENT_TOKEN" | jq -r .accounts[0] 2> /dev/null)
  SUB_SID=$(echo "$ACCOUNT" | jq -r .sid)
  SUB_TOKEN=$(echo "$ACCOUNT" | jq -r .auth_token)
  if [ -z "$SUB_SID" ] || [ "$SUB_SID" = "null" ] || [ -z "$SUB_TOKEN" ] || [ "$SUB_TOKEN" = "null" ]; then
  	echo "{\"sid\": \"false\", \"token\": \"false\"}"
  else
  	echo "{\"sid\": \"$SUB_SID\", \"token\": \"$SUB_TOKEN\"}"
  fi
}

main "$@"
```

Along with a data call in TF to check 

```
provider "twilio" {
  auth_token  = var.auth_token
  account_sid = var.account_sid
}

data "external" "check_subaccount" {
  program  = ["sh", "${path.module}/get_subaccount.sh", var.account_sid, var.auth_token, var.name]
}

provider "twilio" {
  alias        = "sub"
  auth_token   = data.external.check_subaccount.result.token == "false" ? var.auth_token : data.external.check_subaccount.result.token
  account_sid  = data.external.check_subaccount.result.sid == "false" ? var.account_sid : data.external.check_subaccount.result.sid
}
```

**Comment by RJPearson94** [comment]
Hi @enchorb,

Sorry to bring up an old issue but I know you have a workaround for this issue, I created an example and documentation for how you could manage the sub-account and resources in separate state files which would mitigate this issue. 
It uses Terragrunt to manage the link/ dependency between the sub-account and its resources. 

The code and docs can be found at https://github.com/RJPearson94/terraform-provider-twilio/tree/main/examples/account/sub-account.

Hope it helps but if it doesn't then no worries

Kind Regards
Rob

**Comment by RJPearson94** [comment]
Hi @enchorb,

After some more investigation, a new attribute `skip_credential_validation` has been to the provider to allow you to skip credential validation. This should allow you to manage a sub-account and sub-account resources in the same state. I would still recommend using the multiple-state approach but if this is not possible then you have a simpler solution to manage this. An example can be seen below

```hcl
...

provider "twilio" {
  alias       = "sub"
  skip_credential_validation = true
  account_sid = twilio_account_sub_account.sub_account.sid
  auth_token  = twilio_account_sub_account.sub_account.auth_token
}

...
```

You may run into issues if the sub-account is tainted but it appears to work as expected.

This capability has been included in v0.22.0, which is now available in the Terraform Registry.

Please update your Terraform configuration to include the following

```hcl
terraform {
  required_providers {
    twilio = {
      source = "RJPearson94/twilio"
      version = ">= 0.22.0"
    }
  }
} 
```

Hope this helps

Kind Regards
Rob
